### DISCUSSION TITLE
Flatpak release

### DISCUSSION THREAD

**Discussion by mbnoimi** [hardinfo2/hardinfo2#228] [Ideas]
Hi,

Have you considered to make a Flatpak release?

**Comment by hwspeedy** [comment]
Hi Muhammad,

Flatpak/snap is not interesting for hardinfo2.

It is used on small SBC and PC's - and the work is done to support alot of distro for the last 10 years.
Binary packages for download are also ready - but this program should continue/reappear in fedora/debian repos - it has a history and deserves the space.

**Comment by JaneSmith** [comment]
I wish this were available as a Flatpak on Flathub. Flatpak is the native package manager for my distro. It's a pain to install software through other means.

**Comment by hwspeedy** [comment]
Hi @JaneSmith 

> Flatpak is the native package manager for my distro.

Okay, I will take the bait - What distro is that?

**Comment by JaneSmith** [comment]
In my case, it's Fedora Silverblue. It's an immutable/atomic distro where Flatpak is the primary means of installing apps, and it's both discouraged and less convenient to install through other means. There are other Flatpak-centric distros besides Fedora Silverblue, such as Endless OS.

**Comment by hwspeedy** [comment]
@JaneSmith 

The Fedora Atomic distro is rpm-ostree native package manager.

Just do a: "**sudo rpm-ostree install hardinfo2**", reboot and you have hardinfo2 - it will be updated along with the immutable OS (rpm-ostree update, reboot).

We will see if immutable will do something good...

**Comment by mbnoimi** [comment]
I use Linux Mint. It's not available in the main repositories so I've to build it from the source which is really silly (I've 7 PCs)

**Comment by hwspeedy** [comment]
**Ubuntu Packaging wanted**

Yeah, currently nobody with the skills, likes Ubuntu and hardinfo2 enough to put the backports into all maintained Ubuntu repos. 
Hardinfo2 can already be build as package for all maintained Ubuntu versions along with Mint.

See downloads: https://hardinfo2.org/download

**Comment by hwspeedy** [comment]
@mbnoimi

Thanx for info about updated Mint 22.1 distro - I have added it to build system.

See downloads: https://hardinfo2.org/download

Filename: hardinfo2_2.2.6-LinuxMint-22.1_amd64.deb

**Comment by alastortenebris** [comment]
> Hi Muhammad,
> 
> Flatpak/snap is not interesting for hardinfo2.
> 
> It is used on small SBC and PC's - and the work is done to support alot of distro for the last 10 years. Binary packages for download are also ready - but this program should continue/reappear in fedora/debian repos - it has a history and deserves the space.

I don't get this line of logic. Programs don't magically vanish from every distribution's repositories if they're put on Flathub. Mozilla, KDE, and GNOME all have officially maintained flatpaks of their apps, but you won't see distros remove Dolphin just because there's a flatpak of it.

And there are immutable distros that either highly discourage or flat out disable package overlaying. Bazzite, for example, has using rpm-ostree as "not recommended, use as a last resort". And then there's SteamOS. Technically you can unlock the read only file system to install an application, but it'll be wiped upon a system update.

I implore you to reconsider.

**Comment by hwspeedy** [comment]
@alastortenebris 

Please don't complain when the program already works in distros - as many programs this is a hobby. 
If a special distro does not have the hardinfo2 program - that's okay. If you really want it to have the program - do the hacking work - the Linux way. The people before you have paved the way to make this possible.

**Some programs are just not for flatpaks at the current state - kernel, kernel close programs, hardinfo2.**

In Linux eco system - quality is better - each program does the best it can - that's what makes Linux great.

### My five cents:
### Educate
E1. Stop this insane bots everywhere to get flatpak/snap/appimage popular. 
E2. Educate users - problems, what's good, what's bad, how does it work, who does the work, who pays, etc. 
E3. Any technology will do fine if it is the best in what it does.

### Best solution for now
B1. Flatpak should decide whether it is container or binary distribution
B2. Flatpak (same for appimage/snap/etc.) is not verified builds - the source code review work is simply not done - hence the containers half ass solution. Lets see what immutable distros bring...
B3. **Linux is fast because it is optimized** - containers are the other way - copying all different library versions -> so any **distribution updated program is currently the best solution**. Problem: They are only updated for so long and for each distro - that is why RHEL & UBUNTU are a success....

### Distributing binaries is important to Linux but the task has yet to be solved.
D1. Settle on a few golden library versions and make them LTS (20+ years) and create a real access to programs instead of containers could be one solution. 
D2. Allow for non FOSS to be distributed - Linux needs money too and it's needed to get to the famous desktop year - sofar only steam for home gamers and some commercial LTS distros for paid closed source apps.
D3. **How about a common package manager with application maintainers reviewing the code - all in one place for the FOSS packages? - would it kill 100+ distros that are much the same? Would it ruin the great ideas and work from all these distros?**

It's not easy - but do some work for Linux and maybe you could change the world, thanx.

/hwspeedy

**Comment by alastortenebris** [comment]
Boy if that isn't every single point I've heard against flatpak. And I love the "do some work for Linux maybe and you could change the world". But hey, you said if I wanted the program, I could do it "the Linux way".
...so I did that.
https://github.com/alastortenebris/hardinfo2-flatpak

<img width="1283" height="832" alt="Image" src="https://github.com/user-attachments/assets/bcacb372-31a1-4cd5-b718-ec532a2f800c" />

**Comment by hwspeedy** [comment]
@alastortenebris 

I love that you try - this is a good way to move forward - try and then identify problems and fix them.
If you can get to a point where all information is correct, it can be released as hardinfo2 by PR.

Please continue your work and have fun.

Keep us up to speed on your progress and what changes you need to get the information correct.

I have reopened the thread for others to see and join your project: https://github.com/alastortenebris/hardinfo2-flatpak

Good start - now lets get it right!

/hwspeedy

**Comment by roke-julian-lockhart** [comment]
@hwspeedy, I agree with your arguments against containerised solutions, but I don't believe that you've phrased them well, nor do I advise making such arguments unless you can phrase them well.

**Comment by alastortenebris** [comment]
Well, since we're in a discussion now, I'll repost here what does not work currently. Some of this will probably not be able to be fixed due to flatpak sandboxing. We might be able to work around it with sandbox permissions. According to the [Permissions page](https://docs.flatpak.org/en/latest/sandbox-permissions.html), if we allow access to the host, the host files will be mounted at `/run/`. Perhaps that path could be coded in as a search path? 

- OS is listed as Flatpak runtime. 
- hardinfo2 service will most likely not work. I think LACT found a workaround for this, but I couldn't tell you how they did it.
- Groups only show user and nfsnobody.
- Boot log shows nothing
- Memory Devices shows no DMI avaliable.
- Development Environment will not show actual dev environment installed on host system.
- Kernel modules are listed, however any details about them will not show up. lsmod will actually pop up with an error in the command line saying it can't find each module
- File systems shows a lot of bind mounts due to the sandbox.
- A lot of the Network category does not work properly: IP Connections, Routing Table, DNS Servers (Flatpak shows a local dns stub), Statistics, and Shared Directories

The fact that this uses both QT and GTK is a bit of a pain in flatpak land, since they expect you to use one or the other. I found two ways around this: use the GNOME runtime and build qt5-base manually, or use the KDE QT5 runtime, since GTK3 is in the Freedesktop.org runtime that both GNOME and KDE use. GNOME + building QT5 is probably the correct option, but the KDE5 runtime is more optimized.

I will also note that there was an attempt to package hardinfo as a flatpak [here](https://github.com/scx/hardinfo-flatpak). Some of what I wrote in the manifest is based on this.

Some of these modules that have no hope of working properly. Perhaps there could be an option at build time to disable them? If there is one let me know, I can write manifests, but my programming experience is limited to writing some small python scripts for calibre.

**Reply by hwspeedy** [reply]
Good overview - believe a lot more info is wrong/missing. (Try the report compare we talked about)

So a lot of work to get all the information - eg. DBUS service to supply all the missing bits with an installer a la LACT.
Not to speak about the optional programs that users can use on the PC to give additional information, some with USB access like UPS.

**Releasing a downgraded flatpak is not an option as this would impair the project.**

So Sandboxing is not the way for kernel close software...
Would have loved to been proven wrong - and get an easy distribution channel.

A lot can be learned for flatpak if you continue your work of getting all the information correct. But not sure that flatpak want's to go that way anyway - the flathub store is important to project cash.

Try to do an appimage - maybe that is a better binary distribution solution for kernel close software?

**Reply by alastortenebris** [reply]
Yeah, I've never worked with building AppImages before, so I'll be flying blind here. I believe AppImages have access to external programs so it could work? The SystemD service issue is probably still going to be a thing though, but I've never enabled the service in my system package installation of hardinfo2. Honestly, I believe the vast majority of people are either not going to notice or care. Not everyone uses SystemD anyway.

I can try, but It might take me some time.

**Reply by alastortenebris** [reply]
Yeah so it ain't going well.
<img width="1674" height="1081" alt="image" src="https://github.com/user-attachments/assets/16546f2b-5d7b-4031-83f0-89b5e41eaca7" />
I even did the "Screw it just package it from Ubuntu" and it still does this.
Wait, I got something.

**Reply by alastortenebris** [reply]
<img width="1665" height="1072" alt="image" src="https://github.com/user-attachments/assets/a20ab772-da4b-4774-b3e9-2a65c9d4c1a0" />
Well, it's something. It doesn't work on several of my distrobox containers and it crashes if you open the about menu it crashes, but this did work.

Unfortunately, it was done with the most brute-force method possible. Quite literally this is just the Ubuntu package for hardinfo2 packaged as an appimage. I tried building it from the source, but I kept getting what you see above.

By any chance are there hard-coded paths?

**Reply by hwspeedy** [reply]
That's a start with appimage and we don't have to break out of sandbox for everything - we ofcourse want the info from the host and not a container.

**There are some hard coded paths from compile time.** 
library: is found correctly or your would get the first picture.
data: (probably something wrong here - missing background)
locale: translations

crashes are concerning - but could be due pixmaps from data path.

This can be fixed in CMakeList.txt for compile-time paths.



**hardinfo2 -v:**
```
Hardinfo2 version 2.2.17
Copyright (C) 2003-2023 L. A. F. Pereira. 2024-2026 Hardinfo2 Project.

Compile-time options:
  Release version:  Yes (2.2.17) (ARCH_arm)
  LibSoup version:  3.x
  OpenGL version :  QT5
  Vulkan version :  Wayland & X11
  GTK version    :  3.x
  Data           :  /usr/share/hardinfo2
  Library        :  /usr/lib/aarch64-linux-gnu/hardinfo2
  Locale         :  /usr/share/locale
  Compiled for   :  DebianGNULinux-ARCH_arm

```

**Reply by hwspeedy** [reply]
@alastortenebris 

Please share the info for hardinfo2 appimage in another of your git like hardinfo2-appimage - then I will have a look at the missing icons, background (library path), thanx.

**Reply by alastortenebris** [reply]
Unfortunately there really isn't much to share. Essentially I followed the hardinfo2 build instructions and the instructions [here](https://appimage-builder.readthedocs.io/en/latest/intro/tutorial.html).

The issues start to arise once you run the `AppDir/usr/bin/hardinfo2` command (also appimage-builder only really works on Debian 12, but worrying about what to bundle can wait until after the program works. For the record, by test image was a Ubuntu 26.04 distrobox. And yes, I run `sudo apt install lm-sensors sysbench mesa-utils dmidecode udisks2 xdg-utils iperf3 fwupd x11-xserver-utils vulkan-tools gawk`.

Since I couldn't even get to the dependency detection step, I don't have any files to share. I'm afraid you're going to have to compile the program yourself to debug it.

In the hardinfo2 compilation instructions, change the Cmake command to `cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr` and the install command to `make install DESTDIR=AppDir`. Then run `AppDir/usr/bin/hardinfo2`.

**Comment by JaneSmith** [comment]
Why not just break out of the sandbox for low-level tasks? Flatpak and sandboxing are two different things — one is a packaging/distribution model and the other is a security/privacy system. They normally go together (for good reason), but that isn't a hard rule. Apps like GNOME Builder and Ptyxis have their permissions configured such that they're able to escape out of the sandbox and execute commands on the host. As far as I understand it, that solves the issue entirely.

**Reply by hwspeedy** [reply]
Definitely sounds interesting!
The DBUS service was for all "low-level" tasks.
Can we break out of sandbox entirely? or how does it work?
We can change some read /etc/os-release to a cmd - how does it break out?

**Reply by alastortenebris** [reply]
As I mentioned, according to the sandbox permission document, if you give the package the `--host` permission, the host file system should be mounted at `/run`, though Flathub generally discourages doing that. I think it would have to be read only if we want any chance of this being on flathub. It would be a start though.

There's also the `--host-os` permission, but that doesn't give access to `/sys`

**Reply by JaneSmith** [reply]
I'm no expert, but I believe there are various means of escaping the sandbox, one possibly being allowing `--talk-name=org.freedesktop.Flatpak` and then using `flatpak-spawn --host <command>` to execute commands on the host. Apps like this usually show on Flathub with the note "Can acquire arbitrary permissions" to indicate that they can escape the sandbox. I think this approach is quite nice, as it means most of the app can still be safely sandboxed, only poking a hole in it when explicitly needed. Sort of like using "unsafe" in a safe programming language.

**Reply by hwspeedy** [reply]
Discouraged is okay - if the feature is ever removed another thing.

If they cannot be combined - try to give hardinfo2 --host-os and check distro name, system dmi, device tree on arm, sensors, ip connections, etc.

/run is only used for service 3-4 things that need root access - [service](https://github.com/hardinfo2/hardinfo2/blob/master/tools/hwinfo2_fetch_sysdata)
This could be changed - if we get down to just these few being a problem.

**Reply by hwspeedy** [reply]
> flatpak-spawn --host <command>

Could be slow but make a trial reading all modules info or asking the apt cache for info on 50 packages.

**Reply by alastortenebris** [reply]
flatpak-spawn could theoretically work; however, anything that requires root permissions would be a no go. Technically speaking, you theoretically could preface a command with pkexec to get root access, but I don't think that would fly with Flathub.

Edit: Also any command ran on the host would require said program to be installed on the host. So only things like `cat` would work. 

The only other way around it is to trick hardinfo2 to thinking that `/run/host` is actually `/`

To put it simply, anything that requires root access (dmidecode) or membership to the hardinfo2 group is probably a no go. I never actually added myself to that group on my system install anyway, so I don't know what I'm missing.

**Reply by alastortenebris** [reply]
<img width="848" height="570" alt="image" src="https://github.com/user-attachments/assets/2ce568fc-7ef2-4d4b-95a2-9b0b5574e4dd" />
...third times the charm

As a proof of concept, this is being ran in the hardinfo2 flatpak container.

<img width="926" height="245" alt="image" src="https://github.com/user-attachments/assets/7b4b8c91-b3a0-4bc0-a1d3-8091dfad6b9b" />
This is what is pops out if you don't use the `flatpak-spawn --host` command

**Reply by alastortenebris** [reply]
Okay, so the person who attempted to package hardinfo had an idea: create dummy executables for the needed commands and wrap them in a "flatpak-spawn --host" command. Doing this makes "Boots" and most of the network tab work.

```
      - type: script
        commands:
          - flatpak-spawn --host "$( basename "${0}" )" "${@}";
        dest-filename: command.sh
     post-install:
      # install utils
      # Full list: {apcaccess,glxinfo,last,locale,lsb_release,lsmod,lspci,lsusb,modinfo,net,netstat,route,xdpyinfo,xrandr}
      # - glxinfo is provided by mesa-utils module
      # - lspci is provided by pciutils module
      # - xdpyinfo is provided by xdpyinfo module
      # - xrandr is provided by xrandr module
      # Moreover, below we have programs directly in the Freedesktop runtime:
      # - last
      # - locale
      # However, they are useless under the sandbox.
       - echo 'install utils';
         for command in {apcaccess,last,locale,lsb_release,lsmod,lsusb,modinfo,net,netstat,route}; do
         install -p -D -m 0755 "command.sh" "${FLATPAK_DEST}/bin/${command}";
         done;
```

**Reply by hwspeedy** [reply]
Yes, that is doable - we can detect flatpak and prepend paths and flatpak-spawn to commands.

Then there is programs on host machine:
**Hardinfo2 package installs the following packages as dependencies**

Depends:
some changes needed to do inside container pkexec etc. - gawk: Used by hardinfo2 service to determine System Type
should be inside container and access /sys - dmidecode: is needed to provide DMI information.
can be inside container - sysbench: ver 1.0.20 - is needed to run standard sysbench benchmarks.
HMM - needs to be installed - udisks2: is needed to provide storage information.
what works - mesa-utils: glxinfo is needed to get OpenGL info.
what works - lm-sensors: is needed to provide sensors values.
what works - xdg-utils: xdg_open is used to open your browser for bugs, homepage & links.
can be inside container - iperf3: iperf3 is used to benchmark internal network speed.
what works - vulkan-tools: vulkaninfo is used to display vulcan information.
does it work? qt5 or qt6 - qt5-base: QT5 Framework for QT5 OpenGL GPU Benchmark
should work - xcb wayland libdecor-0 : WSI Framework for Vulkan Benchmark
compile time - vulkan glslang-tools : Vulkan Framework/Shader Tool for Vulkan Benchmark
TODO LATER - Service: Service loads SPD modules (at24/ee1004/spd5118) to display SPD info for your DIMMs memory. Show addresses for iomem+ioports.

Recommends/Depends/Optional: (distro choice - prefer installed)
what works - xrandr/x11-xserver-utils: xrandr is used to read monitor setup
how to install - fwupd: fwupd is used to read and display information about firmware in system.

User can install/setup these depending on hardware
should work with flatpak-spawn - apcupsd: apcaccess is used for ups/battery information.
flatpak what works? - mesa-vulkan-swrast/libvulkan_lvp: Vulkan Software driver if you have no hardware vulkan driver (eg. Virtual).
requires a new service - freeipmi-tools/freeipmi: for showing ipmi sensors for Workstations/Servers.
should work - does it? (This is boots) - wtmpdb: for showing latest boots using Who Touched My PC database.

**Reply by alastortenebris** [reply]
It's getting late for me, so I'll end it tonight on this: Several of those are included in the FreeDesktop.org runtime.  Others I have built in the manifest (Mesa/Vulkan Demos, QT5). fwupd and udisks2 just need the DBUS permission and they will work fine.

dmidecode I do have built in the container, but unless it gets root privileges somehow (ideally through polkit), it will not work. Anything dealing with being a member of the "hardinfo2" group is also probably a no go as well.

**Reply by hwspeedy** [reply]
> fwupd and udisks2 just need the DBUS permission

They still need to be installed in host system

dmi decode - can be fetched on new distros using udevadm - if that is hard dependency of flatpak?

**Reply by alastortenebris** [reply]
It's fine for udisks2 and fwupd to need to be on the host system. udisks2 cannot be built into a flatpak anyway. Believe me, I tried. We're fortunate that there's a dbus service for udisks2 otherwise anything that uses it will not work.

udev is not a hard dependency of flatpak, but you will not find a program more ubiquitous that udev (which is a part of SystemD, so every distro with SystemD is going to use it). Even distros that don't use SystemD either use udev (Gentoo) or a eudev, which is a drop-in replacement for it (Alpine). You can safely assume that udev is going to be installed on every distro.

The other device manager is mdev from busybox, but mdev is not recommended for full desktop environments.

**Comment by hwspeedy** [comment]
@alastortenebris 
Please use this repo for our trials: https://github.com/hwspeedy/hardinfo2

When building hardinfo2 add the build flag -DHARDINFO2_FLATPAK=1

1. added flatpak-spawn --host to commands needing sandbox escape (probably needs adjustments)
2. add prefix to files read files: etc, usr, dev, sys - just fixed up os & sensors (Needs adjustments and fix g_test_file, open, g_read_link, etc.
and a lot more.

getting a lot closer to see what is hard to do - we write outstanding issues to your git: https://github.com/alastortenebris/hardinfo2-flatpak

**Reply by alastortenebris** [reply]
Compared to before, lm-sensors is now completly broken. Filesystems also shows incomplete information (I have a secondary drive that it does not detect).

OS Detect does work. I do recommend if possible displaying the details about the installed flatpak runtime (for this build it would be the GNOME Runtime Version 50).

dmidecode also doesn't work, but honestly, I wouldn't even bother since that program needs root access. (It looks like you're planning on switching to udevadm anyway).

The program complains about the xrandr, glxinfo, and vulkaninfo not being installed, but the programs still work as intended.

**Reply by hwspeedy** [reply]
@alastortenebris 

Correct - escaping flatpak half sandbox depending on the interface is a mess, but seams doable for now.

The hardinfo2 program is used on SBC that live with LTS kernel/distros - like Ubuntu 14-26 so we keep the compatibility for ~10 years.

Future looks like udevadm from the dmidecode but still keep 10 years of compatible. So we move forward and keep some compatibility that is slowly removed.

Flatpak used on a distro must be 100% alike with the distro release to keep quality. Except that it will be noted to be flatpak due to lack of compiler optimization being different to continue allowing compare of distros.

Would this sandbox escaping ever be a valid flathub program?

Are we dropping the FLATPAK trials before getting the overview and finding hard stops?

**Reply by alastortenebris** [reply]
As you've stated yourself, a flatpak is a container, so every element that is kept in the container is going to be the same regardless if they're running Debian 12 or Fedora Rawhide.

As for the sandbox, hardinfo2 is a system information tool, and a system information tool needs access to, well... the system. That being said, a worst possible scenario I can see outside of outright denial is forcing the users to manually grant the hardinfo2 flatpak the `--talk-name=org.freedesktop.Flatpak` permission. If that were to happen, I guess you would show a pop-up explaining what to do, similar to how you have pop-ups showing missing programs.

They do allow for apps to break the sandbox; that's why `flatpak-spawn --host` exists, but they won't grant it for every program. Honestly, that is a bridge that we can cross after we get as much working as possible.

**Reply by hwspeedy** [reply]
@alastortenebris 
> As you've stated yourself, a flatpak is a container, so every element that is kept in the container is going to be the same regardless if they're running Debian 12 or Fedora Rawhide.

Nope flatpak sandbox is not emulating a gnome 50 environment, but allowing access to some info sys/proc from system.

> As for the sandbox, hardinfo2 is a system information tool, and a system information tool needs access to, well... the system. That being said, a worst possible scenario I can see outside of outright denial is forcing the users to manually grant the hardinfo2 flatpak the `--talk-name=org.freedesktop.Flatpak` permission. If that were to happen, I guess you would show a pop-up explaining what to do, similar to how you have pop-ups showing missing programs.

That is not going to fly - hardinfo2 needs to work perfectly or not released through flatpak. 
Please checkup by talking to people in the flatpak help group and find out whether it is conflicting.
 
> They do allow for apps to break the sandbox; that's why `flatpak-spawn --host` exists, but they won't grant it for every program. Honestly, that is a bridge that we can cross after we get as much working as possible.

We are not going to release hardinfo2 not working 100% and this needs to figured out as a possible issue.
